data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginPackagesSupplier.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugin;
+
+import com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Properties;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * Supplies packages to scan for plugins.
+ */
+class PluginPackagesSupplier implements Supplier<String[]> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PluginPackagesSupplier.class);
+    private static final String DEFAULT_PLUGINS_CLASSPATH = "com.amazon.dataprepper.plugins";
+
+    PluginPackagesSupplier() {
+    }
+
+    Iterator<URL> loadResources() throws IOException {
+        final Enumeration<URL> resourcesEnumeration = getClass().getClassLoader().getResources("META-INF/data-prepper.plugins.properties");
+        return Iterators.forEnumeration(resourcesEnumeration);
+    }
+
+    @Override
+    public String[] get() {
+
+        final Set<String> packageNames = getUniquePackageNames();
+
+        final String[] packageNamesArray = packageNames.toArray(new String[0]);
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Using packages for plugins: {}", String.join(",", packageNamesArray));
+        }
+
+        return packageNamesArray;
+    }
+
+    private Set<String> getUniquePackageNames() {
+        Set<String> packageNames;
+        try {
+            final Iterator<URL> pluginResources = loadResources();
+            packageNames = readResources(pluginResources);
+        } catch (final IOException ex) {
+            LOG.warn("Unable to load data-prepper.plugins.properties file. Reverting to default plugin package.");
+            packageNames = Collections.singleton(DEFAULT_PLUGINS_CLASSPATH);
+        }
+        return packageNames;
+    }
+
+    private Set<String> readResources(final Iterator<URL> pluginResources) {
+        final Set<String> packageNames = new HashSet<>();
+        while (pluginResources.hasNext()) {
+
+            final URL pluginPropertiesUrl = pluginResources.next();
+
+            final Properties pluginProperties = new Properties();
+            final InputStream pluginPropertiesInputStream;
+            try {
+                pluginPropertiesInputStream = pluginPropertiesUrl.openStream();
+                pluginProperties.load(pluginPropertiesInputStream);
```

Comment:
Help me understand what is happening here a little better. So you add your package name to the end of `META-INF/data-prepper.plugins.properties` file under the `org.opensearch.dataprepper.plugin.packages` variable. Then here we are actually connecting to the package URL and downloading the properties of that package (like package name, and the distribution URL) so we know where to actually go to get the package?

Reply:
This is not downloading anything. Issue #321 discusses downloading jars.

As a developer, you can create your own plugin jar file. If you do, you will also include the `META-INF/data-prepper.plugins.properties` file and specify your plugin package. You can see that in [my sample](https://github.com/dlvenable/data-prepper-sample-plugin/blob/master/src/main/resources/META-INF/data-prepper.plugins.properties#L1).

Then if you run Data Prepper with your plugin in the classpath, Data Prepper will read this file and be able to scan that package from the local jar file.

Now running Data Prepper with a jar in the classpath is a little tricky since we have the uber-jar. You can see how I did it in [the sample](https://github.com/dlvenable/data-prepper-sample-plugin/blob/master/example/run-sample.sh#L5). I expect that #305 will help with this in the future since Data Prepper could include all jars in a `plugins` directory.